### PR TITLE
Add modeling parameter descriptions (#11230)

### LINKED ISSUE #11230: Model parameter description does not exist
### Description
When I try to print the parameter description, I get the following error:
AttributeError: 'Parameter' object has no attribute 'description'

### Expected behavior

According to the documentation, a Parameter object has an attribute description:
https://docs.astropy.org/en/latest/api/astropy.modeling.Parameter.html

### Actual behavior

Output:
```AttributeError: 'Parameter' object has no attribute 'description'```
### Steps to Reproduce

1. Create a model object
2. Print the description of the parameter

```python
from astropy.modeling.powerlaws import PowerLaw1D

func = PowerLaw1D()
print(func.amplitude.description)

```

### System Details

Darwin-18.7.0-x86_64-i386-64bit
Python 3.7.9 | packaged by conda-forge | (default, Dec  9 2020, 20:58:55) 
[Clang 11.0.0 ]
Numpy 1.16.6
astropy 4.0.3
Scipy 1.6.0
Matplotlib 3.3.3

### PR DESCRIPTION
**Pull request by Akshat1Nar**
### Description

This pull request is to address ...

Fixes #11230

**TODO**

- [x] Get okay from maintainers to go forward.
- [x] Add change log under "new feature" under 4.3.
- [x] Add tests.
- [ ] Add user facing documentation.

### CODE REVIEW

**Comment by Akshat1Nar** [comment]
I have submitted this pull request just to get an idea. Basically what I fill be doing to fix this bug is going to every model and adding `description="something"` in every parameter of that model.

**Comment by pllim** [comment]
Hello and thank you! This isn't really a bug, but more of implementing something new, so I adjusted the title, milestone, and so on. I also convert this to draft status for now. FYI.

**Comment by Akshat1Nar** [comment]
Thankyou @pllim , should I continue to fill ```parameter description``` like this or wait for the review?

**Comment by pllim** [comment]
@Akshat1Nar , let's wait a bit so you don't waste your time if there is no buy-in. I added checkboxes above for clarity.

**Comment by hamogu** [comment]
I think this looks good. I'm not a maintainer of this sub-package, but @nden has already agreed that this would be good to have in https://github.com/astropy/astropy/issues/11230#issuecomment-757039065. I think the level of detail it just right. Of course, for the Gaussian it's pretty easy (I mean, what else could a parameter that's called "amplitude" mean, if not the "Amplitude of the Gaussian"?). I think it's too early to pick on individual words, but I wonder if it would be more useful to explain "mean" as something like "position of the peak"? Saying "the parameter called mean is the mean" is true, but not terribly helpful. 

So, I suggest that you go ahead and add descriptions for a few more models. I suggest that we look at it again when you have done may 10-12 models. That way, we have a little more to discuss, but we also make sure that we don't waste your time.

Thanks you so much for helping everyone making astropy better!

**Comment by Akshat1Nar** [comment]
Thankyou @hamogu for that information. I will come back with more descriptions and then you can review again for mistakes. I have also updated first checkbox for now. :)

**Comment by Akshat1Nar** [comment]
@pllim @hamogu  Please review once you are available and help me in proceeding further. Thankyou.

**Review by pllim — COMMENTED** [review]
Implementation-wise, it looks fine, but I am not the person to vet the actual verbiage used.

**Comment by Akshat1Nar** [comment]
> Implementation-wise, it looks fine, but I am not the person to vet the actual verbiage used.

Thankyou @pllim , I will wait for @hamogu 's review then :)

File: astropy/modeling/functional_models.py
Comment (hamogu) [inline astropy/modeling/functional_models.py:2037]:
```suggestion
    amplitude = Parameter(default=1, description="Amplitude (peak) value")
```

**Review by hamogu — CHANGES_REQUESTED** [review]
I left some suggestions. First, thanks for doing this. Please look at my suggestion and comments and see what you think. Please keep in mind that they are just that - suggestions that you can take or leave. You'll also see that in some cases my suggestions differ from what's in the docstring for that model. If you think your or my text for the parameter description is better than what's in there currently, feel free to improve the docstrings as well!
I know from personal experience that this part of the process can be very frustrating with you doing the work and me suggesting something else - and possible another reviewer coming in later who might have other ideas. Having these discussions over small details is how we try to make astropy perfect ;-)

File: astropy/modeling/functional_models.py
Comment (Akshat1Nar) [reply astropy/modeling/functional_models.py:273]: It is mentioned in documentation that angle is in radian.

File: astropy/modeling/functional_models.py
Comment (Akshat1Nar) [reply astropy/modeling/functional_models.py:1685]: At first I did not add it either, but since it was mentioned in Docstring so I thought maybe It had some significance but obiviously not, Sorry.

**Comment by Akshat1Nar** [comment]
@hamogu  I took almost all of the suggestions, they all were necessary. And pushed the changes again. Please review and should I work on models in other files as well. This bug seems to be good-first-bug for me since I am getting to know the codebase as well. :)

File: astropy/modeling/functional_models.py
Comment (hamogu) [reply astropy/modeling/functional_models.py]: There is no need to be sorry. Maybe your wording is better, or maybe my suggestion is better. I'm not sure. That's why I bring this up so that we can discuss and decide together. You don't have to follow all my suggestions, if you don't like them.

And if you and I together decide that there is a better way of saying things, then maybe this is an opportunity to improve the docstring as well!

**Comment by hamogu** [comment]
Looks good. Do you want to continue?

**Comment by Akshat1Nar** [comment]
> Looks good. Do you want to continue?

Yes

**Comment by Akshat1Nar** [comment]
@hamogu Please review, I think I have added description to most of them.

**Comment by Akshat1Nar** [comment]
@hamogu  Review ping.

**Comment by hamogu** [comment]
@Akshat1Nar : Sorry, I have to finish something myself this week. This will have to wait till Friday afternoon or the weekend.

**Comment by Akshat1Nar** [comment]
Thankyou for the response @hamogu  :)

**Comment by pllim** [comment]
@Akshat1Nar , sometimes a reviewer needs a few weeks or even months to circle back. Unfortunately, that is the current reality of FOSS workflow. Thank you for your patience!

**Comment by hamogu** [comment]
@Akshat1Nar  But we should aim to keep it shorter. Months is far too long in my opinion. If it takes that long, please remind!

**Comment by Akshat1Nar** [comment]
Ohh, thankyou for your advice guys, can you also shed some light how can I participate in gsoc or outreachy with `Open Astronomy / Astropy ` . Any IRC channels I need to join any people I need to get familiar with, I am a person with time in hands, I would love to help out :)  (@pllim @hamogu )

**Comment by pllim** [comment]
Re: OpenAstronomy GSOC participation -- There is a `#gsoc` channel on [Astropy Slack](http://joinslack.astropy.org/). 😄

**Comment by Akshat1Nar** [comment]
@pllim Thankyou, I will try to take any upcoming help on gsoc from there :)

File: astropy/modeling/functional_models.py
Comment (hamogu) [inline astropy/modeling/functional_models.py:2222]: This what the description is really good for: `alpha` would not be obvious otherwise. Great!

File: astropy/modeling/physical_models.py
Comment (hamogu) [inline astropy/modeling/physical_models.py:424]: Break line at comma to keep max 80 characters per line?

File: astropy/modeling/projections.py
Comment (hamogu) [inline astropy/modeling/projections.py:220]: Line break: (number of spaces is messed up in github comment, but you get the idea...
```
    gamma = Parameter(default=0.0, getter=_to_orig_unit, setter=_to_radian, 
              description="Look angle γ in degrees (Default = 0°)")
```

Comment (hamogu) [reply astropy/modeling/projections.py:220]: See below., happens a few times that lines are very long. I think it's OK not to break the comment string, but you can certainly break between parameters.

File: astropy/modeling/projections.py
Comment (hamogu) [inline astropy/modeling/projections.py:362]: I don't think we should give defaults here. If you print the parameter, you will see the value. So, I suggest to consistently remove all defaults from all descriptions.

File: astropy/modeling/rotations.py
Comment (hamogu) [inline astropy/modeling/rotations.py:264]:
```suggestion
    lat = Parameter(default=0, getter=_to_orig_unit, setter=_to_radian, description="Longtitude")
```

File: astropy/modeling/rotations.py
Comment (hamogu) [inline astropy/modeling/rotations.py:265]: Longitude of pole

**Review by hamogu — CHANGES_REQUESTED** [review]
A few small comments inline. Otherwise, this just needs a changelog entry. Do you know how to do that?
I don't think tests are needed for this.

**Comment by hamogu** [comment]
Despite what @pllim said above, I don't think any documentation changes are needed. 

However, we do need a test along the line of the example in https://github.com/astropy/astropy/issues/11230 to make sure that all those nice descriptions are actually picked up and can be displayed to the user. (Just test one example. This is really about testing that the Parameters class works.)

**Comment by Akshat1Nar** [comment]
@hamogu Can you please tell me how can I run some of these tests locally, because I am unable to test anything before submitting :(  
linting and all tests

**Comment by hamogu** [comment]
See developer documentation towards the bottom of the page here: https://docs.astropy.org/en/latest/
In particular, tests are covered here: https://docs.astropy.org/en/latest/development/testguide.html

If you can't find anything or it does not work for you, please open an issue so that we can improve the documentation for everyone.

File: astropy/modeling/rotations.py
Comment (dhomeier) [inline astropy/modeling/rotations.py:264]:
```suggestion
    lon = Parameter(default=0, getter=_to_orig_unit, setter=_to_radian, description="Longitude")
    lat = Parameter(default=0, getter=_to_orig_unit, setter=_to_radian, description="Latitude")
```

File: astropy/modeling/rotations.py
Comment (dhomeier) [inline astropy/modeling/rotations.py:225]:
```suggestion
    phi = Parameter(default=0, getter=_to_orig_unit, setter=_to_radian, description="1st Euler angle (Quantity or value in deg)")
    theta = Parameter(default=0, getter=_to_orig_unit, setter=_to_radian, description="2nd Euler angle (Quantity or value in deg)")
    psi = Parameter(default=0, getter=_to_orig_unit, setter=_to_radian, description="3rd Euler angle (Quantity or value in deg)")
```
Note that `Parameter`s can only be numbers, never strings.

File: astropy/modeling/projections.py
Comment (dhomeier) [reply astropy/modeling/projections.py]: Or `print(Model.Parameter.default)`...

File: astropy/modeling/rotations.py
Comment (dhomeier) [inline astropy/modeling/rotations.py:446]:
```suggestion
    angle = Parameter(default=0.0, getter=_to_orig_unit, setter=_to_radian, description="Angle of rotation (Quantity or value in deg)")
```
I _think_ all `Parameter` accept `Quantity` input (the docs e.g. in `projections` are not entirely clear about this), so I'd suggest to follow this style in general – unless it's considered too lengthy.

Comment (hamogu) [reply astropy/modeling/rotations.py:446]: I like @dhomeier suggestion. If the docs are not clear if a parameters accepts a quantity, then @Akshat1Nar can just create a model and try. If the docs and not clear about it, then we should improve the docs by putting that information in the parameter description!

Comment (dhomeier) [reply astropy/modeling/rotations.py:446]: The ones I looked at correctly state `float`; I just was not sure if the docs had not caught up with the wider introduction of quantities. In case of the `Pix2Sky*` etc. models it takes an actual call to `evaluate` to find out that a quantity raises, so it may be safe enough to assume the current docstrings are correct...
Generally it would certainly be preferable if all models could handle quantities of matching type, but that's a project of its own.

**Comment by Akshat1Nar** [comment]
Thankyou @dhomeier and @hamogu . I will create a perfectly working pull request by today.

File: astropy/modeling/functional_models.py
Comment (dhomeier) [inline astropy/modeling/functional_models.py:1053]:
```suggestion
    amplitude_L = Parameter(default=1, description=
                            "Lorentzian amplitude (peak value of the associated Lorentz function)")
```
Note that I have updated some of the docstrings for the Gauss and Lorentz-related functions in #11177; if the `description` strings here are in sync, I can merge them when the time has come.

**Comment by pllim** [comment]
Lots of unrelated commits.

**Comment by Akshat1Nar** [comment]
@pllim  I was trying to rebase this onto the current head and this happened, any idea how to fix this?

**Comment by hamogu** [comment]
Do you remember what you did when you tried to rebase? That might help me understand what went wrong.

**Comment by dhomeier** [comment]
Looks like you rebased in the wrong direction, has happened to me on my first attempts as well.
Can you reset your local branch to 66eaa30, which should be the last commit before the rebase attempt (make sure not to lose any changes you might have added after this)?

**Comment by Akshat1Nar** [comment]
@dhomeier  I requested information on Astropy Slack's `#general` channel. Not to spam the comments with doubts, Please help me there. : )

**Comment by Akshat1Nar** [comment]
@dhomeier   Do I need to add tests for every parameter description of every model? I have also added a changelog entry by looking at documentation please review.

**Comment by hamogu** [comment]
No, we don't need a separate test for every parameter description, but one test that loops over a few of them, just to make sure it works in general.

File: astropy/modeling/projections.py
Comment (dhomeier) [inline astropy/modeling/projections.py:801]:
```suggestion
    mu = Parameter(default=1.0, description="Distance from center of sphere in spherical radii")
```
I think this may safely be shortened a bit, otherwise the docstring itself would need to be broken into two lines.
But perhaps someone familiar with that model could better comment on what is a sufficient description.

**Review by dhomeier — CHANGES_REQUESTED** [review]
Just a few style/formatting remarks for today; in general we try to avoid over-/underindentation (E128) even though it is not checked in the flake test.
I recommend to set up your editor or syntax checker locally as described in https://docs.astropy.org/en/stable/development/codeguide.html#coding-style-conventions to have this formatting done mostly automatically.

**Comment by hamogu** [comment]
Almost done! I have nothing to add to @dhomeier suggestions. I think when you implement them (which can be as easy as pressing the "commit suggestion" button under his comments!) then you are done! Thank you so much for staying on this and helping us to make astropy better in big and small ways!

**Comment by Akshat1Nar** [comment]
> Almost done! I have nothing to add to @dhomeier suggestions. I think when you implement them (which can be as easy as pressing the "commit suggestion" button under his comments!) then you are done! Thank you so much for staying on this and helping us to make astropy better in big and small ways!

Thankyou very much @hamogu I committed the suggested changes.

File: astropy/modeling/functional_models.py
Comment (dhomeier) [inline astropy/modeling/functional_models.py:1076]:
```suggestion
    amplitude_L = Parameter(default=1,     # noqa: N815
                            description="The Lorentzian amplitude")
    fwhm_L = Parameter(default=2/np.pi,    # noqa: N815
                       description="The Lorentzian full width at half maximum")
    fwhm_G = Parameter(default=np.log(2),  # noqa: N815
                       description="The Gaussian full width at half maximum")
```
What I meant in my earlier comment was that the `# noqa: N815` tells the syntax checker to ignore the uppercase in the variable names, so it needs to be on the line where they are defined (it is perhaps a bit perfectionist, since those warnings are not treated as failures in the code style checks, so one could simply ignore them; but it makes it easier to keep track of other syntax violations that might remain). So except for `amplitude_L` everything still fits on a single line, so it would have been easier to keep it that way, but when breaking the line, it needs to be done like above. Also note the 2 blanks minimum space before inline comments.

**Review by dhomeier — CHANGES_REQUESTED** [review]

**Comment by dhomeier** [comment]
Thanks for the changes and sorry I haven't had time to fully review the content yet, since we have a coordination committee meeting this week.
The changelog entry should be under the appropriate `astropy.modeling` subsection. Also, since this was labelled as an enhancement (and an enhancement it definitely is!), it probably should not go under "Bug Fixes", but into "New Features" or perhaps "API Changes" – @pllim what do you think?

**Comment by eteq** [comment]
Since this hasn't made the feature freeze cutoff I'm re-milestoning to 5.0

**Comment by hamogu** [comment]
Uuups. I forgot about that.

**Comment by hamogu** [comment]
Unfortunately, one change is that we''re now using town-crier instead of the Changelog file: https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#add-a-changelog-entry

Since that's really me and @dhomeier that let it linger for so long, I can fix that up when everything else is done.

**Comment by hamogu** [comment]
The other conflict is east, and I can also take care of that. What this means is, that there is no need to rebase right now, we can review and see if any more changes to the content are required and then re-base at the very end.

**Comment by Akshat1Nar** [comment]
Hello @hamogu . What do I need to do to help to complete this pull request?

**Comment by dhomeier** [comment]
Must have missed the update from 1419388 here; apart from merging the test this looks good, but will probably need another review after rebasing, given the extensive documentation changes from #11118 and #11595.

**Comment by hamogu** [comment]
@Akshat1Nar Do you want to try the doc changes or do you want me to do that? It's kind of a special situation of a PR that was opened before we switched to towncrier and so needs to be updated to that.

Instructions are here: https://github.com/astropy/astropy/pull/11279#issuecomment-813398923
(except that "master" is now called "main", another one time change that just happens to fall on this).

If you want to do that, you are welcome to, if not, I'll do it in the next few days (adding on top of your commits so that you still get full credit for what you did)

**Comment by hamogu** [comment]
OK, I changed to the CHANGES to the towncrier format and force-pushed that on top of @Akshat1Nar changes. Assuming that the CI still passes, I think this is ready for final review (which is why I added the "ready-for-final-review" label ;-)

**Comment by pllim** [comment]
@hamogu , does this mean the two reviews that requested changes can be dismissed? Thanks!

**Comment by hamogu** [comment]
@dhomeier asked to review again after this, and I'll look through it, too. However, this is a case of "perfect is the enemy of good". The new description is better then what we had before (nothing), so cases where we've looked at that and decided that it's no ideal, but it mirrors the docstring should, in my opinion, not hold this up. Of course, I'd like to hear from @nden if we can.
(And I guess my review is now not good enough to merge any longer, because I'm one of the authors now ;-)

**Review by hamogu — APPROVED** [review]

**Comment by hamogu** [comment]
@dhomeier I believe that all the changes you requested were already done by @Akshat1Nar , but github did not update the display; not sure why. I've checked and manually marked them as resolved.

**Review by dhomeier — APPROVED** [review]

**Comment by dhomeier** [comment]
Thanks, yes I believe they had been marked resolved earlier this week, but maybe got lost again in the rebase.
Any syncing to descriptions from #11595 can be done in a follow-up.

**Pull request merged**
